### PR TITLE
Multiple API fixes in UART driver

### PR DESCRIPTION
**Pull request by playfulFence**
closes https://github.com/esp-rs/esp-hal/issues/2733
closes https://github.com/esp-rs/esp-hal/issues/2726
closes https://github.com/esp-rs/esp-hal/issues/2724
closes https://github.com/esp-rs/esp-hal/issues/2636 (public methods only, post your thoughts here if you want private functions to follow these conventions too)

### CODE REVIEW

File: esp-hal/src/uart.rs [hunk 1]
```diff
@@ -511,6 +512,8 @@ impl Default for Config {
     }
 }
 
+#[derive(procmacros::BuilderLite)]
+#[non_exhaustive]
```
Comment (bugadani) [inline esp-hal/src/uart.rs:516]: At the same time we can mark this as `#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]` + defmt::Format as well, I think.

File: esp-hal/src/uart.rs [hunk 2]
```diff
@@ -528,25 +531,14 @@ pub struct AtCmdConfig {
     pub char_num: Option<u8>,
```
Comment (bjoernQ) [inline esp-hal/src/uart.rs:531]: I wonder if we should make this `Option<core::num::NonZeroU8>` - `Some(0)` doesn't make sense and if `None` we use `1`

Would save us two bytes - not sure if it's worth it - makes things a bit less nice in user code

Comment (bugadani) [reply esp-hal/src/uart.rs:531]: This can also just be `u8`, can't it?

File: esp-hal/src/uart.rs [hunk 3]
```diff
@@ -1582,7 +1575,15 @@ where
     }
 
     fn flush(&mut self) -> Result<(), Self::Error> {
-        self.tx.flush()
+        loop {
```
Comment (bjoernQ) [inline esp-hal/src/uart.rs:1578]: `embedded_io::Write::flush(&mut self.tx)` would safe us from duplicating this code

File: esp-hal/src/uart.rs (hunk 2 above)
Comment (bjoernQ) [reply esp-hal/src/uart.rs:531]: yes - sounds good

Comment (bjoernQ) [reply esp-hal/src/uart.rs:531]: (probably needs some explanation in the doc-comment)

Comment (playfulFence) [reply esp-hal/src/uart.rs:531]: Then we need to validate `AtCmdConfig` if `char_num` is not 0 which is related https://github.com/esp-rs/esp-hal/issues/2839 and probably should not be done in this PR (?)

Comment (bugadani) [reply esp-hal/src/uart.rs]: 0 might be a perfectly reasonable value of "I don't want to detect AT commands"

Comment (bjoernQ) [reply esp-hal/src/uart.rs]: `0` just means the same as currently `None` - shouldn't change anything?

BUT I just realized that some chips (e.g. ESP32 at least) offer more than 16 bits for the counts and the time-out.
Big question before stabilization would be if we are fine with artificially limiting this or not

Comment (bjoernQ) [reply esp-hal/src/uart.rs]: oh and I think it solves a problem: currently we cannot undo the at-cmd-config once set? After the change the user could set `char_num` to `0` and effectivly disable it

**Review by bugadani — APPROVED** [review]
Maybe the constructor change should be in the migration guide.

**Review by bjoernQ — APPROVED** [review]
LGTM

**Pull request merged**